### PR TITLE
Some assertions are not including variable names in assertion message

### LINKED ISSUE #1353: Some assertions are not including variable names in assertion message
### Description

When using fluent assertions for things like objects & lists, assertion failures will typically log the names of the variables. This doesn't appear to happen when comparing `DateTime`s, `Guid`s, objects, etc.

It is very useful to have the variables names for context, as when the assertion message doesn't have these it is difficultto figure out which assertion failed when looking at test output for "AssertionScope" blocks (and for non-assertion scope blocks, you can only find the failing assertion via the line numbers in the stack).

Here is what the output looks like for one of the assertions which includes variables names:
```csharp
var z = "SSSS";
z.Should().Contain("BBB");

> Expected z "SSSS" to contain "bbb".
```

### Complete minimal example reproducing the issue

```csharp
var x = DateTime.UtcNow;
var y = DateTime.UtcNow.AddSeconds(1);

x.Should().Be(y);
```

### Expected behavior:

Assertion failure should be something like: `Expected x to be <2020-07-15 16:51:43.1621939>, but was <2020-07-15 16:51:43.5991913>.`

### Actual behavior:

`Expected date and time to be <2020-07-15 16:51:43.1621939>, but found <2020-07-15 16:51:43.5991913>.`

### Versions

* FluentAssertions 5.10.2
* .NET framework 4.8

### PR DESCRIPTION
**Pull request by Serg046**
## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by a new or existing set of unit tests which follow the Arrange-Act-Assert syntax such as is used [in this example](https://github.com/fluentassertions/fluentassertions/blob/daaf35b9b59b622c96d0c034e8972a020b2bee55/Tests/FluentAssertions.Shared.Specs/BasicEquivalencySpecs.cs#L33).
* [x] If the contribution adds a feature or fixes a bug, please update [**the release notes**](https://github.com/fluentassertions/fluentassertions/tree/master/docs), which are published on the [website](https://fluentassertions.com/releases).
* [x] If the contribution changes the public API the changes needs to be included by running `AcceptApiChanges.ps1`/`AcceptApiChanges.sh`.
* [ ] If the contribution affects [the documentation](https://github.com/fluentassertions/fluentassertions/tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

Fixes #1353

### CODE REVIEW

File: Src/FluentAssertions/CallerStatementBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                if (isQuoteContext)
+                {
+                    statement.Append(symbol);
+                }
+
+                previousChar = symbol;
+                return isQuoteContext ? Result.Handled : Result.InProgress;
+            }
+
+            private bool IsAtEscaped()
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerStatementBuilder.cs:96]: 🔧 As mentioned before, I would rename this to `IsVerbatim`

File: Src/FluentAssertions/CallerStatementBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            {
+                if (previousChar == '@')
+                {
+                    return true;
+                }
+
+                if (statement.Length > 1)
+                {
+                    var idx = statement.Length - 1;
+                    return statement[idx--] == '$'
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerStatementBuilder.cs:106]: ❌ Please don't mix increments like these. People always forget in what order things happen.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
That's quite an impressive PR. To be fair, I would not have invested the time to support all the misuse of the language in your examples. But that's the awesomeness of the community ❤

**Comment by dennisdoomen** [comment]
@Serg046 did you give up on this?

**Comment by Serg046** [comment]
I am just waiting for the holidays. If you still think we could merge something like this, I will complete it soon.

**Comment by dennisdoomen** [comment]
@jnyrup what do you think?

**Comment by jnyrup** [comment]
I think it looks quite interesting. I just haven't taken the time to digest this fully.

I was mostly concerned about other PRs keeping caller identification lazy.

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 3]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            // Act
+            Action act = () => foo.Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo to be <null>*");
+        }
+
+        [Fact]
+        public void When_variable_is_not_captured_it_should_use_the_variable_name()
```
[the remark this answers is not in the document]
Comment (Serg046) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:70]: It was added when I started with Mono.Cecil. That time such test was important, now it is not. I will just remove this test and maybe similar ones

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 4]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            // Act
+            Action act = () => foo.Bar.Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo.Bar to be <null>*");
+        }
+
+        [Fact]
+        public void When_method_name_contains_get_it_should_not_remove_the_prefix()
```
[the remark this answers is not in the document]
Comment (Serg046) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:114]: The whole test could be removed because the current code base just rely on symbols, no need to distinguish properties and methods

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 5]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            {
+                Field = "test3"
+            };
+
+            // Act
+            Action act = () => foo.GetFoo(test1).GetFooStatic("test" + 2).GetFoo(foo.Field).Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo.GetFoo(test1).GetFooStatic(\"test\"+2).GetFoo(foo.Field) to be <null>*");
```
[the remark this answers is not in the document]
Comment (Serg046) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:156]: I think it is not that hard. We could look for the last dot which is not part of arguments list. The reason why I didn't do it like you suggest is that I saw the current behaviour better. Just let me know if you think it is worth doing. I can do it right here (guess it is better) or implement it as a separate PR.

File: Src/FluentAssertions/CallerStatementBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                    result = handlerEnumerator.Current.Handle(symbolEnumerator.Current);
+                }
+            }
+
+            return result;
+        }
+
+        public override string ToString() => statement.ToString();
+
+        public enum Result
```
[the remark this answers is not in the document]
Comment (Serg046) [reply Src/FluentAssertions/CallerStatementBuilder.cs:44]: Yes, I remember that I was not happy by this thing. The answer is in this method:
```csharp
        public Result Append(string symbols)
        {
            Result result = Result.InProgress;
            using var symbolEnumerator = symbols.GetEnumerator();
            while (symbolEnumerator.MoveNext() && result != Result.Done)
            {
                result = Result.InProgress;
                using var handlerEnumerator = handlers.GetEnumerator();
                while (handlerEnumerator.MoveNext() && result == Result.InProgress)
                {
                    result = handlerEnumerator.Current.Handle(symbolEnumerator.Current);
                }
            }

            return result;
        }
```
I have 2 cycles and 2 flags for exit. I will try to refactor it somehow, anyway I need to move some things as suggested above.

**Comment by rygwdn** [comment]
Is there still interest in this PR? I would love to see #1353 fixed!

**Comment by dennisdoomen** [comment]
> Is there still interest in this PR? I would love to see #1353 fixed!

If somebody is willing to finalize the PR, then yes.

File: Src/FluentAssertions/CallerStatementBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            Handled,
+            Done
+        }
+
+        private interface IHandler
+        {
+            Result Handle(char symbol);
+        }
+
+        private class QuotesHandler : IHandler
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs:56]: Done

File: Src/FluentAssertions/CallerIdentifier.cs [hunk 8]
```diff
@@ -133,6 +125,25 @@ private static string GetSourceCodeLineFrom(StackFrame frame)
...
+        private static string GetSourceCodeStatementFrom(StackFrame frame, StreamReader reader, string line)
+        {
+            int column = frame.GetFileColumnNumber();
+            if (column > 0)
+            {
+                line = line.Substring(Math.Min(column - 1, line.Length - 1));
+            }
+
+            var sb = new CallerStatementBuilder();
+            CallerStatementBuilder.Result state;
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentifier.cs:137]: Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 1 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs:96]: Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 2 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs:106]: Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs:44]: @Serg046 Did you have any thoughts/suggestions you could share on this? I don't mind making the changes, there's just a lot going on 😛

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:70]: I've removed this one - but I'm not sure what the similar ones that can be removed are

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 9]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+                foo.Should().BeNull();
+            };
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo to be <null>*");
+        }
+
+        [Fact]
+        public void When_field_is_the_caller_it_should_use_the_field_name()
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:85]: Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Serg046) [reply Src/FluentAssertions/CallerStatementBuilder.cs:44]: @mdentremont, the only suggestion is to refactor it somehow. I just remember that processing was line by line where Handled means that the line is processed and we need to switch next.

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:156]: @dennisdoomen I would be OK with leaving this as is as long as you are - I kind of like the extra context (although if it were too long via chained method calls, it could we unwieldy - then again with EF this could end up showing `.ToList()` which wouldn't be helpful).

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 10]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            // Act
+            Action act = () => foo.get_BarMethod("test\";").Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo.get_BarMethod(\"test\\\";\") to be <null>*");
+        }
+
+        [Fact]
+        public void When_parameters_contain_at_escaped_quote_it_should_include_that_to_the_caller()
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:221]: Done

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 11]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            // Act
+            Action act = () => foo.get_BarMethod(@"test"";").Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo.get_BarMethod(@\"test\"\";\") to be <null>*");
+        }
+
+        [Fact]
+        public void When_parameters_contain_at_and_dollar_escaped_quote_it_should_include_that_to_the_caller()
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:235]: Done

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs [hunk 12]
```diff
@@ -50,6 +51,296 @@ public void When_variable_name_contains_Should_it_should_identify_the_entire_var
...
+            // Act
+            Action act = () => foo.get_BarMethod(@$"test"";").Should().BeNull();
+
+            // Assert
+            act.Should().Throw<XunitException>()
+                .WithMessage("*Expected foo.get_BarMethod(@$\"test\"\";\") to be <null>*");
+        }
+
+        [Fact]
+        public void When_parameters_contain_comments_it_should_include_them_to_the_caller()
```
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs:249]: I think it's because the string has contents which could be comments in C# (`//` and `/**/`) - I'll reword the title a bit

File: Src/FluentAssertions/CallerStatementBuilder.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Collections.Generic;
+using System.Text;
+using FluentAssertions.CallerIdentification;
+
+namespace FluentAssertions
+{
+    internal class CallerStatementBuilder
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerStatementBuilder.cs:7]: 🔧 I would move this file to the `CallerIdentification` folder as well. It's an implementation detail of the `CallerIdentifier`

File: Src/FluentAssertions/CallerIdentification/IHandler.cs [hunk 14]
```diff
@@ -0,0 +1,7 @@
+﻿namespace FluentAssertions.CallerIdentification
+{
+    public interface IHandler
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerIdentification/IHandler.cs:3]: 🔧 Thinking about the algorithm, it seems to me that this is really represents a parsing strategy, so I suggest to rename this to `IParser` or `IParsingStrategy`
🔧 I would also add some docs and explain its purpose, especially about the fact that these handlers are state-full and created once per parsing task.

File: Src/FluentAssertions/CallerIdentification/HandlerResult.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+﻿namespace FluentAssertions.CallerIdentification
+{
+    public enum HandlerResult
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerIdentification/HandlerResult.cs:3]: 🤔 Isn't this the _state_ of the parsing process? If so, I think renaming this to `ParsingState` or something like that makes the algorithm a lot easier to grasp.

File: Src/FluentAssertions/CallerIdentification/HandlerResult.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+    {
+        InProgress,
+        Handled,
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerIdentification/HandlerResult.cs:6]: 🤔 I don't quite understand the difference between `Handled` and `Done`. Maybe the name isn't quite right, or some docs can help here.

File: Src/FluentAssertions/CallerIdentification/QuotesHandler.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+            previousChar = symbol;
+            return isQuoteContext ? HandlerResult.Handled : HandlerResult.InProgress;
+        }
+
+        private bool IsVerbatim()
+        {
+            return previousChar == '@'
+               || (statement.Length > 1
+                   && statement[statement.Length - 1] == '$'
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerIdentification/QuotesHandler.cs:49]: 🔧 Isn't it clearer to use `EndsWith("$@")` or use the new C# range syntax?

File: Src/FluentAssertions/CallerIdentification/ShouldCallHandler.cs [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        private const string ShouldCall = ".Should";
+        private readonly StringBuilder statement;
+
+        public ShouldCallHandler(StringBuilder statement) => this.statement = statement;
+
+        public HandlerResult Handle(char symbol)
+        {
+            if (statement.Length >= ShouldCall.Length)
+            {
+                var ldx = statement.Length - 1;
```
Comment (dennisdoomen) [inline Src/FluentAssertions/CallerIdentification/ShouldCallHandler.cs:16]: 🤔 What's `ldx` and `rdx`?

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
👍🏻 Impressive piece of work
❌ Please make everything related to caller identification `internal`.

File: Src/FluentAssertions/CallerIdentification/QuotesHandler.cs [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+            if (isQuoteContext)
+            {
+                statement.Append(symbol);
+            }
+
+            previousChar = symbol;
+            return isQuoteContext ? HandlerResult.Handled : HandlerResult.InProgress;
+        }
+
+        private bool IsVerbatim()
```
Comment (jnyrup) [inline Src/FluentAssertions/CallerIdentification/QuotesHandler.cs:45]: `@$""` is also a valid interpolated verbatim string.

File: docs/_pages/releases.md [hunk 20]
```diff
@@ -39,6 +39,7 @@ Changes since 6.0.0 Alpha 2
 * Better parameter checking of `MethodInfoSelectorAssertions` - [#1569](https://github.com/fluentassertions/fluentassertions/pull/1569)
 * Better parameter checking of `XDocumentAssertions`, `XElementAssertions` and `XAttributeAssertions` - [#1564](https://github.com/fluentassertions/fluentassertions/pull/1564)
 * In a chained assertion API call, a second call to `ForCondition` should not even evaluate its lambda when the previous assertion failed - [#1587](https://github.com/fluentassertions/fluentassertions/pull/1587)
+* Improved caller name determination by supporting multiple lines, comments and semicolons - [#1435](https://github.com/fluentassertions/fluentassertions/pull/1435).
```
Comment (jnyrup) [inline docs/_pages/releases.md:42]: Beta 1 has shipped, this will be part of `6.0.0 xxx`

File: Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mdentremont) [reply Tests/FluentAssertions.Specs/Execution/CallerIdentifierSpecs.cs]: Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs]: Ah no worries. I don't think it's all that bad as it stands - it takes a second to wrap your head around, but then it's decently straightforward.

If you have any refactoring ideas you'd like to apply, maybe just make them in a future PR 👍

**Comment by mdentremont** [comment]
> 
> 
> 👍🏻 Impressive piece of work
> ❌ Please make everything related to caller identification `internal`.

Done

File: Src/FluentAssertions/CallerStatementBuilder.cs (hunk 13 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerStatementBuilder.cs:7]: Done

File: Src/FluentAssertions/CallerIdentification/IHandler.cs (hunk 14 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/IHandler.cs:3]: Done

File: Src/FluentAssertions/CallerIdentification/HandlerResult.cs (hunk 15 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/HandlerResult.cs:3]: Done

File: Src/FluentAssertions/CallerIdentification/HandlerResult.cs (hunk 16 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/HandlerResult.cs:6]: After coming back to this code again, I think maybe `GoToNextSymbol` makes a lot more sense.

File: Src/FluentAssertions/CallerIdentification/QuotesHandler.cs (hunk 17 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/QuotesHandler.cs:49]: Both of those suggestions sound great, but 1 requires changing the framework, and the other (`EndsWith` doesn't work on `StringBuilder`, and calling `ToString` kind of defeats the purpose).

Maybe we leave as is?

File: Src/FluentAssertions/CallerIdentification/ShouldCallHandler.cs (hunk 18 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/ShouldCallHandler.cs:16]: I assume it's something like "left index" & "right index" - I'll rename it accordingly, hopefully that's good 🤞

File: Src/FluentAssertions/CallerIdentification/QuotesHandler.cs (hunk 19 above)
Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/QuotesHandler.cs:45]: I think that's what it is doing?

File: docs/_pages/releases.md (hunk 20 above)
Comment (mdentremont) [reply docs/_pages/releases.md:42]: Done

**Comment by mdentremont** [comment]
@dennisdoomen @jnyrup Thanks so much for the feedback you two!

While addressing the feedback, I noticed that the tests that did `WithMessage("*...*")` were being passed even though the parts inside the wildcards were not in the message (this sounds like a possible `WithMessage` bug?).

I also noticed some edge cases while reading through the code, so I added new tests to confirm my suspicions (the tests failed).

I've made a decent amount of refactoring, and have also fixed all of the broken tests (and of course addressed as much of the review feedback as I could, and replied where I couldn't).

Cheers!

**Review by dennisdoomen — APPROVED** [review]
Some unit tests are failing...

File: Src/FluentAssertions/CallerIdentification/QuotesHandler.cs (hunk 19 above)
Comment (jnyrup) [reply Src/FluentAssertions/CallerIdentification/QuotesHandler.cs]: Ahh sorry, I meant that `$@""` is also a valid interpolated verbatim string.

Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/QuotesHandler.cs]: ahh OK - I'll try to get this resolved today hopefully.

Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/QuotesHandler.cs]: Done

File: Src/FluentAssertions/CallerIdentification/ShouldCallParsingStrategy.cs [hunk 21]
```diff
@@ -4,7 +4,7 @@ namespace FluentAssertions.CallerIdentification
 {
     internal class ShouldCallParsingStrategy : IParsingStrategy
     {
-        private const string ShouldCall = ".Should()";
+        private const string ShouldCall = ".Should(";
```
Comment (jnyrup) [inline Src/FluentAssertions/CallerIdentification/ShouldCallParsingStrategy.cs:7]: Is there any disadvantage to using `.Should(` over `.Should()`?
I imagine `Should` is only called with an argument in our tests when we fake timing.

Comment (mdentremont) [reply Src/FluentAssertions/CallerIdentification/ShouldCallParsingStrategy.cs:7]: It _think_ it seems reasonable  this way it also still works if somebody were to overload `.Should` via an extension method (not sure why they would though).

**Review by jnyrup — APPROVED** [review]
